lib/services/role.go

```diff
@@ -3288,14 +3291,56 @@ func (set RoleSet) GetAllowedSearchAsRoles() []string {
 	}
 	for _, role := range set {
 		for _, a := range role.GetSearchAsRoles(types.Allow) {
-			if _, ok := denied[a]; !ok {
+			filterResult := true
+			if len(allowFilters) > 0 {
+				for _, filter := range allowFilters {
+					if !filter(role) {
+						filterResult = false
+						break
+					}
+				}
+			}
+			if _, ok := denied[a]; !ok && filterResult {
 				allowed = append(allowed, a)
 			}
```

Comment:
to go with my other suggestion
```suggestion
			if _, isDenied := denied[a]; isDenied {
				continue
			}
			allowed = append(allowed, a)
```